### PR TITLE
PR: Fix unchecking `Display variables and attributes` not updating tree (Outline explorer)

### LINKED ISSUE #21456: Unchecking Display variables and attributes in outline does not work
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

Unchecking the "Display variables and attributes" option in the outline pane does not work until Spyder is restarted.

### What steps reproduce the problem?

1. Check the "Display variables and attributes" option in the outline pane
2. Uncheck the "Display variables and attributes" option in the outline pane
3. The variables and attributes are still displayed in the outline pane

### What is the expected output? What do you see instead?

Unchecking the "Display variables and attributes" option should work as expected.

### Paste Traceback/Error Below (if applicable)

No error/trace message.

## Versions

* Spyder version: 5.4.5  (conda)
* Python version: 3.11.6 64-bit
* Qt version: 5.15.8
* PyQt5 version: 5.15.9
* Operating System: macOS-13.6-arm64-arm-64bit

### Dependencies

```
# Mandatory:
applaunchservices >=0.3.0                                                   :  0.3.0 (OK)
atomicwrites >=1.2.0                                                        :  1.4.1 (OK)
chardet >=2.0.0                                                             :  5.2.0 (OK)
cloudpickle >=0.5.0                                                         :  3.0.0 (OK)
cookiecutter >=1.6.0                                                        :  2.4.0 (OK)
diff_match_patch >=20181111                                                 :  20230430 (OK)
intervaltree >=3.0.2                                                        :  3.1.0 (OK)
IPython >=7.31.1,<9.0.0,!=8.8.0,!=8.9.0,!=8.10.0,!=8.11.0,!=8.12.0,!=8.12.1 :  8.16.1 (OK)
jedi >=0.17.2,<0.19.0                                                       :  0.18.2 (OK)
jellyfish >=0.7                                                             :  1.0.1 (OK)
jsonschema >=3.2.0                                                          :  4.19.1 (OK)
keyring >=17.0.0                                                            :  24.2.0 (OK)
nbconvert >=4.0                                                             :  7.9.2 (OK)
numpydoc >=0.6.0                                                            :  1.5.0 (OK)
parso >=0.7.0,<0.9.0                                                        :  0.8.3 (OK)
pexpect >=4.4.0                                                             :  4.8.0 (OK)
pickleshare >=0.4                                                           :  0.7.5 (OK)
psutil >=5.3                                                                :  5.9.5 (OK)
pygments >=2.0                                                              :  2.16.1 (OK)
pylint >=2.5.0,<3.0                                                         :  2.17.7 (OK)
pylint_venv >=3.0.2                                                         :  3.0.3 (OK)
pyls_spyder >=0.4.0                                                         :  0.4.0 (OK)
pylsp >=1.7.4,<1.8.0                                                        :  1.7.4 (OK)
pylsp_black >=1.2.0,<3.0.0                                                  :  1.3.0 (OK)
qdarkstyle >=3.0.2,<3.2.0                                                   :  3.1 (OK)
qstylizer >=0.2.2                                                           :  0.2.2 (OK)
qtawesome >=1.2.1                                                           :  1.2.3 (OK)
qtconsole >=5.4.2,<5.5.0                                                    :  5.4.4 (OK)
qtpy >=2.1.0                                                                :  2.4.1 (OK)
rtree >=0.9.7                                                               :  1.1.0 (OK)
setuptools >=49.6.0                                                         :  68.2.2 (OK)
sphinx >=0.6.6                                                              :  7.2.6 (OK)
spyder_kernels >=2.4.4,<2.5.0                                               :  2.4.4 (OK)
textdistance >=4.2.0                                                        :  4.5.0 (OK)
three_merge >=0.1.1                                                         :  0.1.1 (OK)
watchdog >=0.10.3                                                           :  3.0.0 (OK)
zmq >=22.1.0                                                                :  25.1.1 (OK)

# Optional:
cython >=0.21                                                               :  None (NOK)
matplotlib >=3.0.0                                                          :  None (NOK)
numpy >=1.7                                                                 :  1.26.0 (OK)
pandas >=1.1.1                                                              :  2.1.1 (OK)
scipy >=0.17.0                                                              :  None (NOK)
sympy >=0.7.3                                                               :  None (NOK)
```

### PR DESCRIPTION
**Pull request by remisalmon**
## Description of Changes

This fixes https://github.com/spyder-ide/spyder/issues/21456 where unchecking "Display variables and attributes" in the outline does not work.

Issue debug log:
```
2023-10-31 11:28:00,930 [DEBUG] [spyder.plugins.editor.widgets.editor] -> Refresh EditorStack
2023-10-31 11:28:00,931 [DEBUG] [spyder.plugins.outlineexplorer.widgets] -> Set current editor to file /Users/remisalmon/test.py
2023-10-31 11:28:00,940 [DEBUG] [spyder.config.manager] -> Sending notification to observers of display_variables option in configuration section outline_explorer
2023-10-31 11:28:00,940 [DEBUG] [spyder.plugins.outlineexplorer.widgets] -> Current and new trees for file /Users/remisalmon/test.py are the same, so no update is necessary
```

I believe this uses the wrong operator from the `intervaltree` package where the difference operator does work as expected in this case:

with a file like:
```python
def function():
    return

x = 0
```
we get, after unchecking Display variables and attributes:
```
current_tree=IntervalTree([Interval(0, 3, ((0, 2), function, 0b7cc8e0-5ff3-4535-b9e9-aa01e97ba571, False)), Interval(3, 4, ((3, 3), x, deb51838-ac1c-457a-b0ab-ce8449fd1e14, False))])
tree=IntervalTree([Interval(0, 3, ((0, 2), function, bbd41a82-41b7-460f-b56f-d5fca0f1e6f8, False))])
tree - current_tree = IntervalTree()
tree == current_tree = False
```
(the difference returns an empty tree object but the trees are different)

Another fix would be to look at `changes = current_tree - tree` but we do not need the actual difference tree object `changes` so this makes the code simpler.

This also fixes an issue where the tree would not update when "Display variables and attributes" is checked and a variable or attribute is removed from the code.

Tested locally.

### Issue(s) Resolved

Fixes #21456

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: remisamon

### CODE REVIEW

File: spyder/plugins/outlineexplorer/widgets.py [hunk 1]
```diff
@@ -608,8 +608,7 @@ def update_tree(self, items, editor):
         if not must_update:
             # Compare with current tree to check if it's necessary to update
             # it.
-            changes = tree - current_tree
-            if tree and len(changes) == 0:
+            if tree and tree == current_tree:
```
Comment (remisalmon) [inline spyder/plugins/outlineexplorer/widgets.py:611]: I am not sure if we still need to check that the tree is not empty?

Comment (dalthviz) [reply spyder/plugins/outlineexplorer/widgets.py:611]: I think is not needed anymore :+1:

**Review by dalthviz — COMMENTED** [review]
Thank you @remisalmon for helping checking for a fix! I think is okay to remove the `tree` being nonempty validation. Also, one missing thing is to add a test for this. Maybe the test could use the code example you used to illustrate the update issue, what do you think?

If you need help with the test let us know and again thank you for helping us with this!

**Comment by remisalmon** [comment]
Ignore the pep8speaks message I force pushed a fix but it is not going away :)

I added a test that is passing locally (log below).

I could not figure out how to get qtbot to click on the `DisplayVariables` action that is not a button so I just forced it with `toggle_variables()` but maybe not ideal.

```
> py.test spyder/plugins/outlineexplorer/tests/test_widgets.py::test_display_variables
================================================= test session starts =================================================
platform linux -- Python 3.9.18, pytest-6.2.5, py-1.11.0, pluggy-1.3.0
PyQt5 5.15.9 -- Qt runtime 5.15.8 -- Qt compiled 5.15.8
rootdir: /home/remi/Work/GitHub/spyder, configfile: pytest.ini
plugins: flaky-3.7.0, cov-4.1.0, lazy-fixture-0.6.3, mock-3.12.0, order-1.0.1, qt-4.2.0, timeout-2.2.0
collected 1 item

spyder/plugins/outlineexplorer/tests/test_widgets.py .                                                          [100%]

================================================== warnings summary ===================================================
spyder/utils/programs.py:27
  /home/remi/Work/GitHub/spyder/spyder/utils/programs.py:27: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

../../Miniforge3/envs/spyder-dev/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
  /home/remi/Work/Miniforge3/envs/spyder-dev/lib/python3.9/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../Miniforge3/envs/spyder-dev/lib/python3.9/site-packages/jupyter_client/connect.py:22
  /home/remi/Work/Miniforge3/envs/spyder-dev/lib/python3.9/site-packages/jupyter_client/connect.py:22: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write

-- Docs: https://docs.pytest.org/en/stable/warnings.html
===Flaky Test Report===

test_display_variables passed 1 out of the required 1 times. Success!

===End Flaky Test Report===
============================================ 1 passed, 3 warnings in 1.34s ============================================
```

File: spyder/plugins/outlineexplorer/widgets.py (hunk 1 above)
Comment (remisalmon) [reply spyder/plugins/outlineexplorer/widgets.py]: Done in https://github.com/spyder-ide/spyder/pull/21481/commits/c3606519d3f5015a4be5123ee72c624c91b44e53.

**Comment by dalthviz** [comment]
Awesome @remisalmon ! I think is okay to use the `toggle_variables` directly and don't worry about the pep8 message (usually we don't enforce them if they are only complaining about tests files). Could you remove the `TODOs` regarding the usage of `qtbot` please? Other than that this LGTM :+1:

**Comment by remisalmon** [comment]
Should be all good now, thank you @dalthviz for reviewing this.

**Review by dalthviz — APPROVED** [review]
Thanks @remisalmon !

**Pull request merged**